2DRove/Assets/Scripts/StateMachines/SpitterFSM/SpitterAttackState.cs

```diff
@@ -1,34 +1,149 @@
+    private SpitterStateManager spitter;
+
+    public float attackCooldown = 2.0f; // Cooldown between attacks
+    private float rangedAttackDistance = 10.0f; // The distance for ranged attacks
+    private bool isAttackInitiated = false; // To control attack initiation
+    private Coroutine attackCoroutine; // To keep track of the coroutine
+
     public override void EnterState(SpitterStateManager spitter)
     {
-        Debug.Log("Spitter Entering Attack State");
-        attackTime =0.9f; 
+        this.spitter = spitter;
+        playerTransform = GameObject.FindGameObjectWithTag("Player")?.transform; // Use null-conditional to prevent error
         animator = spitter.GetComponent<Animator>();
-        animator.SetBool("attacking", true);
+
+        if (playerTransform != null)
+        {
+            isAttackInitiated = false; // Reset attack initiation flag
+            animator.SetBool("isAttacking", true);
+            attackCoroutine = spitter.StartCoroutine(AttackRoutine()); // Store coroutine reference
+        }
+        else
+        {
+            spitter.SwitchState(spitter.IdleState);
+        }
     }
 
     public override void UpdateState(SpitterStateManager spitter)
     {
-        if (attackTime <= 0)
+        // If the player is destroyed during the attack, clean up and exit state
+        if (playerTransform == null || playerTransform.gameObject == null)
         {
-            spitter.SwitchState(spitter.IdleState);
-            animator.SetBool("attacking", false);
+            CleanupAttack();
+            return;
+        }
+
+        float distanceToPlayer = Vector3.Distance(spitter.transform.position, playerTransform.position);
+        if (!isAttackInitiated && distanceToPlayer <= rangedAttackDistance)
+        {
+            attackCoroutine = spitter.StartCoroutine(AttackRoutine());
+        }
+        else if (distanceToPlayer > rangedAttackDistance && isAttackInitiated)
+        {
+            CleanupAttack();
         }
-        attackTime -= Time.deltaTime;
+    }
+
+    private IEnumerator AttackRoutine()
+    {
+        isAttackInitiated = true;
+        yield return new WaitForSeconds(attackCooldown); // Wait for the appropriate time to fire the projectile
+
+        // If player has been destroyed, don't continue
+        if (playerTransform == null)
+        {
+            yield break;
+        }
+
+        FireProjectile(); // Proceed to fire the projectile
+
+        yield return new WaitForSeconds(attackCooldown); // Cooldown after firing
+
+        // If player has been destroyed, don't continue
+        if (playerTransform == null || playerTransform.gameObject == null)
+        {
+            CleanupAttack();
+
+            yield break;
+        }
+
+        CleanupAttack();
+    }
+
+    private void FireProjectile()
+    {
+        if (playerTransform == null) return; // Extra check to ensure player is not null
+
+        GameObject projectilePrefab = spitter.ProjectilePrefab; 
```

Comment:
What prefab is placed in here? This is where the projectile that the main character has should be. I did see that you had the particles for the projectile, which is nice, but add the man projectile as well.

Reply:
The range projectile will be place here.

Reply:
Alright, then we'll leave it as for now, and make changes to the implementation later, thank you.